### PR TITLE
feat: max pool voting power

### PR DESCRIPTION
**Pull request by troykessler**
## Release Notes

- **New Features**
  - Introduced maximum voting power per pool, enhancing the staking mechanism.
  - Added a new error for pools with zero delegation.

- **Bug Fixes**
  - Added test cases for various scenarios, including effective stake evaluations and pool validations.

- **Documentation**
  - Updated CHANGELOG.md to reflect new features and bug fixes.

- **Tests**
  - Enhanced test coverage for effective stakes and voting power constraints across multiple scenarios.
  - Added tests for invalid bundles and scenarios involving stake fractions and foreign delegations.

### CODE REVIEW

File: x/stakers/keeper/exported_functions.go [hunk 1]
```diff
@@ -124,21 +124,131 @@ func (k Keeper) GetValidatorPoolCommission(ctx sdk.Context, staker string, poolI
...
+			Address: address,
+			Stake:   stake,
+		})
+		totalStake += int64(stake)
+	}
+
+	totalStakeRemainder := totalStake
+
+	// sort descending based on stake
+	sort.Slice(validators, func(i, j int) bool {
```
Comment (mbreithecker) [inline x/stakers/keeper/exported_functions.go]: Should probably use StableSort

File: x/stakers/keeper/exported_functions.go [hunk 2]
```diff
@@ -79,29 +81,27 @@
...
 
-		if delegation > highestDelegation {
-			highestDelegation = delegation
-		}
+	if maxVotingPower.IsZero() {
+		return true
 	}
 
-	return totalDelegation, highestDelegation
+	return math.LegacyOneDec().Quo(maxVotingPower).GT(math.LegacyNewDec(addresses))
```
[the remark this answers is not in the document]
Comment (troykessler) [reply x/stakers/keeper/exported_functions.go:104]: There is no potential division by zero if the addresses are empty

File: x/stakers/keeper/exported_functions.go [hunk 3]
```diff
@@ -194,17 +323,27 @@
...
-	slashFraction := k.getSlashFraction(ctx, slashType).Mul(k.GetValidatorPoolStakeFraction(ctx, staker, poolId))
+	// here the stake fraction can be actually different from the stake fraction the validator has specified
+	// for this pool because with the max voting power in place his effective stake in a pool could have been
+	// reduced because his original stake was too high. Therefore, we determine the true stake fraction
+	// by dividing his effective stake with his bonded amount.
+	stakeFraction := math.LegacyZeroDec()
+
+	if !validator.GetBondedTokens().IsZero() {
+		stakeFraction = math.LegacyNewDec(int64(k.GetValidatorPoolStake(ctx, staker, poolId))).QuoInt(validator.GetBondedTokens())
+	}
```
[the remark this answers is not in the document]
Comment (troykessler) [reply x/stakers/keeper/exported_functions.go:342]: The effective stake can not be greater than the bonded tokens of the validator. This is ensured by the `GetValidatorPoolStakes` method

File: x/stakers/keeper/exported_functions.go [hunk 4]
```diff
@@ -124,21 +124,150 @@ func (k Keeper) GetValidatorPoolCommission(ctx sdk.Context, staker string, poolI
...
+			stakes[validator.Address] -= uint64(cutoffAmount)
+
+			// we take the cutoff amount and distribute it on the remaining validators down the list
+			// who all have less voting power than the current one. We distribute the cutoff amount
+			// based on the validator's stake
+			if totalStakeRemainder > 0 {
+				for _, v := range validators[i+1:] {
+					stakes[v.Address] += uint64(math.LegacyNewDec(int64(v.Stake)).QuoInt64(totalStakeRemainder).MulInt64(cutoffAmount).TruncateInt64())
+				}
+			}
```
[the remark this answers is not in the document]
Comment (troykessler) [reply x/stakers/keeper/exported_functions.go:224]: There are not enough $KYVE in existence to trigger an integer overflow

**Review by mbreithecker — APPROVED** [review]

**Pull request merged**